src/lib/data/attachment.ts

```diff
@@ -0,0 +1,344 @@
+import {
+  calculateReflectiveTransitiveEntitySupport,
+  ValidatedPlanEntities,
+} from './planEntities';
+
+const ATTACHMENT_DATA = t.union([
+  t.type({
+    type: t.literal('caseLoad'),
+    value: ATTACHMENT_VERSION_VALUE['caseLoad'],
+  }),
+  t.type({
+    type: t.literal('contact'),
+    value: ATTACHMENT_VERSION_VALUE['contact'],
+  }),
+  t.type({
+    type: t.literal('cost'),
+    value: ATTACHMENT_VERSION_VALUE['cost'],
+  }),
+  t.type({
+    type: t.literal('fileWebContent'),
+    value: ATTACHMENT_VERSION_VALUE['fileWebContent'],
+  }),
+  t.type({
+    type: t.literal('indicator'),
+    value: ATTACHMENT_VERSION_VALUE['indicator'],
+  }),
+  t.type({
+    type: t.literal('textWebContent'),
+    value: ATTACHMENT_VERSION_VALUE['textWebContent'],
+  }),
+]);
+
+type PlanAttachmentData = t.TypeOf<typeof ATTACHMENT_DATA>;
+
+type PlanAttachmentParent =
+  | { type: 'plan' }
+  | { type: 'governingEntity'; id: GoverningEntityId }
+  | { type: 'planEntity'; id: PlanEntityId };
+
+export type PlanAttachmentResult = {
+  id: AttachmentId;
+  customRef: string;
+  /**
+   * The part of the plan under which this attachment is directly nested
+   */
+  parent: PlanAttachmentParent;
+  /**
+   * If set, this particular plan attachment is associated with the given
+   * governing entity
+   */
+  governingEntity: GoverningEntityId | null;
+  /**
+   * List of all plan entities that this attachment is associated with,
+   * along with any entities that are transitively supported
+   */
+  supportsEntities: PlanEntityId[];
+  /**
+   * The attachment type, along with the type-checked data from its value
+   */
+  data: PlanAttachmentData;
+};
+
+type PlanAttachmentResults = Map<AttachmentId, PlanAttachmentResult>;
+
+/**
+ * Get all of the attachments of a particular type for a particular plan.
+ *
+ * In addition to fetching the attachments, this also validates the contents of
+ * their value field, ensuring they are of the correct type.
+ */
+export const getAllPlanAttachments = async ({
+  database,
+  planId,
+  types,
+  planEntities,
+  governingEntities,
+  log,
+}: {
+  database: Database;
+  planId: PlanId;
+  types: AttachmentType[];
+  version: 'latest';
+  /**
+   * A map of plan entities, which **must** include any entity that is the
+   * parent of the given attachment.
+   *
+   * Usually all the entities for a given plan are provided
+   */
+  planEntities: ValidatedPlanEntities;
+  /**
+   * A map of governing entities, which **must** include any governing entity
+   * that is the parent of the given attachment.
+   *
+   * Usually all the governing entities for a given plan are provided
+   */
+  governingEntities: MapOfGoverningEntities;
+  log: SharedLogContext;
+}): Promise<PlanAttachmentResults> => {
+  const prototypes = await database.attachmentPrototype.find({
+    where: (builder) => builder.whereIn('type', types),
+  });
+  const attachmentPrototypes = organizeObjectsByUniqueProperty(
```

Comment:
`attachmentPrototypes` is more appropriate for what is called `prototypes`
```suggestion
  const attachmentPrototypesById = organizeObjectsByUniqueProperty(
```